File: exercises/cpp/parser_code_review/expr.cpp
#include "expr.h"
#include <cmath>
#include <stdexcept>

static char op_symbol(OpType op) {
    switch (op) {
        case OpType::Sum:      return '+';
        case OpType::Mult:     return '*';
        case OpType::Div:      return '/';
        case OpType::Exponent: return '^';
    }
    throw std::logic_error("unknown op");
}

static int int_pow(int base, int exp) {
    if (exp < 0) throw std::domain_error("negative exponent not supported");
    int result = 1;
    for (int i = 0; i < exp; ++i)
        result *= base;
    return result;
}

int NumberExpr::evaluate() const {
    return value;
}

std::string NumberExpr::format() const {
    if (value < 0)
        return "(" + std::to_string(value) + ")";
    return std::to_string(value);
}

int BinaryOpExpr::evaluate() const {
    int l = left->evaluate();
    int r = right->evaluate();
    switch (op) {
        case OpType::Sum:      return l + r;
        case OpType::Mult:     return l * r;
        case OpType::Div:
            if (r == 0) throw std::domain_error("division by zero");
            return l / r;
        case OpType::Exponent: return int_pow(l, r);
    }
    throw std::logic_error("unknown op");
}

std::string BinaryOpExpr::format() const {
    return "(" + left->format() + " " + op_symbol(op) + " " + right->format() + ")";
}

File: exercises/csharp/parser-code-review/ExpressionEvaluator/Expr.cs
namespace ExpressionEvaluator;

public enum OpType { Sum, Mult, Div, Exponent }

public abstract record Expr
{
    public abstract int Evaluate();
    public abstract string Format();
}

public sealed record NumberExpr(int Value) : Expr
{
    public override int Evaluate() => Value;

    public override string Format() =>
        Value < 0 ? $"({Value})" : Value.ToString();
}

public sealed record BinaryOpExpr(OpType Op, Expr Left, Expr Right) : Expr
{
    public override int Evaluate()
    {
        int l = Left.Evaluate();
        int r = Right.Evaluate();
        return Op switch
        {
            OpType.Sum => l + r,
            OpType.Mult => l * r,
            OpType.Div => r == 0
                ? throw new InvalidOperationException("division by zero")
                : l / r,
            OpType.Exponent => r < 0
                ? throw new InvalidOperationException("negative exponent not supported")
                : IntPow(l, r),
            _ => throw new InvalidOperationException($"unknown op: {Op}")
        };
    }

    public override string Format()
    {
        char symbol = Op switch
        {
            OpType.Sum => '+',
            OpType.Mult => '*',
            OpType.Div => '/',
            OpType.Exponent => '^',
            _ => throw new InvalidOperationException($"unknown op: {Op}")
        };
        return $"({Left.Format()} {symbol} {Right.Format()})";
    }

    private static int IntPow(int @base, int exp)
    {
        int result = 1;
        for (int i = 0; i < exp; i++)
            result *= @base;
        return result;
    }
}
